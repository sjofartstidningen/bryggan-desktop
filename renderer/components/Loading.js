import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import { Zap } from './Icon';

const LoadingWrapper = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  padding-top: 1rem;
  color: ${p => p.theme.color.darkGrey};
`;

const IconWrapper = styled.div`
  margin-bottom: 0.5rem;
`;

const MessageWrapper = styled.p`
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica,
    Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol',
    sans-serif;
  font-size: 0.75rem;
  width: auto;
  max-width: 150px;
  text-align: center;
`;

function Loading({ threshold, message }) {
  const [showLoader, setShowLoader] = useState(() => threshold < 1);

  useEffect(() => {
    if (threshold > 0) {
      const timeoutId = setTimeout(() => setShowLoader(true), threshold);
      return () => clearTimeout(timeoutId);
    }
  }, []);

  return showLoader ? (
    <LoadingWrapper>
      <IconWrapper>
        <Zap />
      </IconWrapper>
      <MessageWrapper>{message}</MessageWrapper>
    </LoadingWrapper>
  ) : null;
}

Loading.defaultProps = {
  threshold: 0,
  message: 'Loading',
};

export { Loading };
