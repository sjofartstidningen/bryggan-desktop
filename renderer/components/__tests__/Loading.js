import React from 'react';
import { render, flushEffects, waitForElement } from 'react-testing-library';
import { Loading } from '../Loading';

describe('Component: <Loading />', () => {
  it('should render a loader', () => {
    const { container, getByText } = render(<Loading />);
    expect(getByText(/loading/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should be able to render only after a threshold', async () => {
    const { getByText } = render(<Loading threshold={10} />);

    expect(() => getByText(/loading/i)).toThrow();

    flushEffects();
    await waitForElement(() => getByText(/loading/i));

    expect(getByText(/loading/i)).toBeInTheDocument();
  });

  it('should accept a message to supply the loading indicator', () => {
    const message = 'Hello I am loading';
    const { getByText } = render(<Loading message={message} />);
    expect(getByText(message)).toBeInTheDocument();
  });
});
