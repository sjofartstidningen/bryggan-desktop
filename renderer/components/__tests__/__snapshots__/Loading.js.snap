// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <Loading /> should render a loader 1`] = `
.c2 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  width: 1em;
  height: 1em;
}

.c3 {
  width: 1em;
  height: 1em;
  fill: none;
  stroke: currentColor;
  stroke-width: 2;
  stroke-linecap: round;
  stroke-linejoin: round;
  position: absolute;
  bottom: -0.125em;
}

.c0 {
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-top: 1rem;
  color: rgb(68,68,68);
}

.c1 {
  margin-bottom: 0.5rem;
}

.c4 {
  font-family: -apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Helvetica,Arial,sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol',sans-serif;
  font-size: 0.75rem;
  width: auto;
  max-width: 150px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        <svg
          class="c3"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            points="13 2 3 14 12 14 11 22 21 10 12 10 13 2"
          />
        </svg>
      </span>
    </div>
    <p
      class="c4"
    >
      Loading
    </p>
  </div>
</div>
`;
